src/Installer/Microsoft.Dotnet.Installation/Internal/DotnetArchiveExtractor.cs

```
@@ -56,68 +54,143 @@ public void Prepare()
+            {
+                // If we can't rename, just continue - extraction will overwrite
+            }
+        }
+
+        // Step 3: Extract the archive (all files directly since muxer has been renamed)
         if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
         {
-            return ExtractTarArchive(archivePath, targetDir, muxerConfig, installTask);
+            ExtractTarArchive(archivePath, targetDir, installTask);
         }
         else
         {
-            return ExtractZipArchive(archivePath, targetDir, muxerConfig, installTask);
+            ExtractZipArchive(archivePath, targetDir, installTask);
+        }
+
+        // Step 4: If there was a previous muxer, compare versions and restore if needed
+        if (hadExistingMuxer && File.Exists(muxerTempPath))
+        {
+            Version? newMuxerVersion = GetLatestRuntimeVersionFromInstallRoot(targetDir);
+
+            // If the new version is less than the old version, restore the old muxer
+            if (newMuxerVersion != null && existingMuxerVersion != null && newMuxerVersion < existingMuxerVersion)
+            {
+                try
+                {
+                    if (File.Exists(muxerTargetPath))
+                    {
+                        File.Delete(muxerTargetPath);
+                    }
+                    File.Move(muxerTempPath, muxerTargetPath);
+                }
+                catch
+                {
+                    // If we can't restore, the new muxer will remain
+                }
+            }
+            else
+            {
+                // New version is >= old version, or we couldn't determine versions - keep new muxer
+                try
+                {
+                    if (File.Exists(muxerTempPath))
+                    {
+                        File.Delete(muxerTempPath);
+                    }
+                }
+                catch
+                {
+                    // Ignore cleanup errors
```

Let's also throw an exception here.